### PR TITLE
OCPBUGS-3505: Waiting for 2 masters in HA mode case

### PR DESCRIPTION
**Pull request by tsorya**
OCPBUGS-3505: Waiting for 2 masters in HA mode case.
This is needed to handle the case where one master is ready and all
needed pods are running though the second one didn't join yet and
keepalived can't move vip yet.

### CODE REVIEW

File: pkg/start/start.go [hunk 1]
```diff
@@ -135,6 +138,19 @@ func (b *startCommand) Run() error {
...
 	}
+
+	sno, err := isSingleNodeTopology(b.assetDir)
+	if err != nil {
+		UserOutput(err.Error())
+		return err
+	}
+	if !sno {
+		UserOutput("Waiting for 2 ready masters")
+		if err = waitUntilMastersAreReady(ctx, client, 2); err != nil {
```
Comment (eranco74) [inline pkg/start/start.go:149]: Do we really need to wait until they are ready?
it's not enough if they join?

File: pkg/start/start.go [hunk 2]
```diff
@@ -244,3 +260,44 @@ func makeBootstrapSuccessEvent(ns, name string) *corev1.Event {
 	}
 	return event
 }
+
+
+func isSingleNodeTopology(assetDir string) (bool, error) {
```
Comment (eranco74) [inline pkg/start/start.go:265]: Consider renaming to 
```suggestion
func isSingleNodeControlPlane(assetDir string) (bool, error) {
```

File: pkg/start/status.go [hunk 3]
```diff
@@ -32,6 +32,67 @@ func waitUntilPodsRunning(ctx context.Context, c kubernetes.Interface, pods map[
 	return nil
 }
 
+func waitUntilMastersAreReady(ctx context.Context, c kubernetes.Interface, requireNumOfMasters int) error {
+	sc, err := newNodeStatusController(c, requireNumOfMasters)
```
Comment (eranco74) [inline pkg/start/status.go:36]:
```suggestion
	nsc, err := newNodeStatusController(c, requireNumOfMasters)
```

File: pkg/start/start.go (hunk 1 above)
Comment (tsorya) [reply pkg/start/start.go:149]: IMO why not to wait a little bit more to be sure both are fine? Though we can have only one ready and one joined

Comment (cybertron) [reply pkg/start/start.go:149]: Strictly speaking we just need two nodes joined so when one of them comes up it will have a peer for keepalived.

Comment (tsorya) [reply pkg/start/start.go]: NP, fixed it to wait for 2 masters to join

**Review by eranco74 — COMMENTED** [review]
Looks good

File: pkg/dependencymagent/doc.go [hunk 4]
```diff
@@ -1,3 +1,4 @@
+//go:build tools
```
Comment (mfojtik) [inline pkg/dependencymagent/doc.go:1]: why this change?

File: pkg/start/status.go [hunk 5]
```diff
@@ -32,6 +32,67 @@ func waitUntilPodsRunning(ctx context.Context, c kubernetes.Interface, pods map[
...
+	requireNumOfMasters int
+	joinedNames         map[string]string
+}
+
+func newNodeStatusController(client kubernetes.Interface, requireNumOfMasters int) (*nodesStatusController, error) {
+	return &nodesStatusController{client: client, requireNumOfMasters: requireNumOfMasters, joinedNames: map[string]string{}}, nil
+}
+
+func (s *nodesStatusController) Run() {
+	options := metav1.ListOptions{LabelSelector: "node-role.kubernetes.io/master"}
```
Comment (mfojtik) [inline pkg/start/status.go:62]: not sure if this is the right label for master node...

@openshift/openshift-team-node

File: cmd/cluster-bootstrap/start.go [hunk 6]
```diff
@@ -51,6 +52,7 @@ func init() {
 	cmdStart.Flags().DurationVar(&startOpts.terminationTimeout, "tear-down-termination-timeout", 0, "wait of (graceful) termination of the bootstrap control-plane before reporting success. Set to zero to disable.")
 	cmdStart.Flags().DurationVar(&startOpts.tearDownDelay, "tear-down-delay", 0, "duration to delay the bootstrap control-plane tear-down before bootstrap-success event is created, in order to give load-balancers time to observe the self-hosted control-plane. This even applies in case of --tear-down-early.")
 	cmdStart.Flags().DurationVar(&startOpts.assetsCreatedTimeout, "assets-create-timeout", time.Duration(60)*time.Minute, "how long to wait for all the assets be created.")
+	cmdStart.Flags().IntVar(&startOpts.requiredNumberOfJoinedMaster, "required-num-masters-joined", 2, "how many masters should join before the start command does the pivot, used only for ha installation")
```
[the remark this answers is not in the document]
Comment (tsorya) [reply cmd/cluster-bootstrap/start.go:55]: In order to allow changing if needed, not sure that it should be hardcoded.

File: pkg/start/status.go (hunk 5 above)
Comment (tsorya) [reply pkg/start/status.go:62]: Yeah it is.
Was already tested e2e by assisted qe

File: pkg/dependencymagent/doc.go (hunk 4 above)
Comment (tsorya) [reply pkg/dependencymagent/doc.go:1]: something automatic, missed this file

File: pkg/start/status.go (hunk 5 above)
Comment (rphillips) [reply pkg/start/status.go:62]: The master is node-role is correct.

File: cmd/cluster-bootstrap/start.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (dgrisonnet) [reply cmd/cluster-bootstrap/start.go:55]: If we don't have any use-cases for this option I would prefer if we just hardcode it since it will be harder to remove the option in the future if needed since it could break some users.

**Review by dgrisonnet — COMMENTED** [review]
could you please add a unit-test for the new NodeStatusController?

**Comment by tsorya** [comment]
> could you please add a unit-test for the new NodeStatusController?

I would say there is not much to write unitest for. It only checks number of nodes in the list. The only testable func is RequiredNumberOfMastersJoined and it is pretty straightforward without any complicated logic

File: pkg/start/status_test.go [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+				Labels: map[string]string{"node-role.kubernetes.io/master": ""},
+			}}
+		nsController.client.CoreV1().Nodes().Create(context.TODO(), node, metav1.CreateOptions{})
+		// wait for cache to sync, if cache sync will fail for unknown reason, test will fail as no host will be added
+		for i := 0; i < 8; i++ {
+			if nsController.nodeController.HasSynced() {
+				break
+			}
+			time.Sleep(250 * time.Millisecond)
+		}
```
Comment (dgrisonnet) [inline pkg/start/status_test.go:29]: This looks like something that could become flaky.

How about providing the node objects in the test cases and create them by passing them to `fakeclient.NewSimpleClientset()`?

Comment (tsorya) [reply pkg/start/status_test.go:29]: You can't, the problem is that code is using cache and it is feeling itself around 1 sec after. Creating host and providing it to fakeclient is totally the same as fakeclient just return the object you added though cache should sync itself and it takes sometime

File: pkg/start/status.go [hunk 8]
```diff
@@ -32,6 +32,68 @@ func waitUntilPodsRunning(ctx context.Context, c kubernetes.Interface, pods map[
...
+			ListFunc: func(lo metav1.ListOptions) (runtime.Object, error) {
+				return s.client.CoreV1().Nodes().List(context.TODO(), options)
+			},
+			WatchFunc: func(lo metav1.ListOptions) (watch.Interface, error) {
+				return s.client.CoreV1().Nodes().Watch(context.TODO(), options)
+			},
+		},
+		&v1.Node{},
+		1*time.Millisecond,
+		//30*time.Minute,
```
Comment (dgrisonnet) [inline pkg/start/status.go:75]: we should stick to 30 minutes here

**Review by dgrisonnet — COMMENTED** [review]
/lgtm
/approve
